### DISCUSSION TITLE
Why are the url parameters to an api parsed correctly by this lib but not by others?

### DISCUSSION THREAD

**Discussion by riha** [Xunnamius/next-test-api-route-handler#519] [Q&A] [answered]
I'm new to Next.js and I have previously tested my API endpoint using [supertest](https://www.npmjs.com/package/supertest). That worked fine until I tried to get the parsed URL parameters in an API - just keep getting `undefined` in the parameters property of the request in the API handler. So (in for example and API route `/api/test/[id]` the id when requesting `/api/test/33` `33` won't be parsed out to the parameters object in the API handler).

Same happened then using [next-test-api-route-handler](https://github.com/Xunnamius/next-test-api-route-handler) until I specifically set the `params` object in `testApiHandler`. Then it worked 😅

I'm just trying to understand what's actually going on. What's different from calling the API in a test from actually doing the request from the frontend? Is there some middleware in next that isn't run when running the test?

**Comment by Xunnamius** [comment, marked as answer]
Your instincts are correct that Next.js is invoking some dynamic parameter parsing "middleware" [somewhere else](https://github.com/vercel/next.js/blob/f16ee05f599de27e777ac2b736c3bf820a19bd7b/packages/next/server/base-server.ts#L499-L504) that is _not_ invoked when just running the API resolver function in isolation, and certainly not when emulating Next.js routing via a request testing library (supertest, node-mocks-http, etc). In fact, Next.js's API resolver function uses [dependency injection](https://github.com/vercel/next.js/blob/f16ee05f599de27e777ac2b736c3bf820a19bd7b/packages/next/server/api-utils/node.ts#L161) to get its URL parameters object, which NTARH [takes full advantage of](https://github.com/Xunnamius/next-test-api-route-handler/blob/b7e6811e2c055852a56c2c76947280f0cadaa74a/src/index.ts#L211) to inject the `params` object during testing. This is why setting `params` in NTARH just works 🙂.

[This answer](https://github.com/Xunnamius/next-test-api-route-handler/issues/303#issuecomment-903344572) might shed more light on things too.
